### PR TITLE
Bug 1876211 - Add MDN for Suggest [firefox-android: main]

### PR DESCRIPTION
**Pull request by ncloudioj**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

First stab, will need to rebase against #6085. Also noticed a few duplicate code paths that require more refactors perhaps in this patch.

### CODE REVIEW

**Comment by ncloudioj** [comment]
@0c0w3 wanted to confirm with you that the current implementation only returns the raw data from RS, do we want to process the data here? (e.g. add utm params; add the hardcoded icon URI, etc.)

**Review by 0c0w3 — APPROVED** [review]
Thanks @ncloudioj, this looks great! Sorry for the delay, I would have r+'ed this earlier today but I wanted to do an end-to-end test in Firefox and that took some time. It works great.

Yes, the current impl only reflects the data in RS. The desktop urlbar JS handles UTM params and the icon. (Maybe they should be handled by Rust too but that's a future discussion.)

**Comment by ncloudioj** [comment]
Thanks for confirming that, Drew!

File: components/suggest/src/db.rs [hunk 1]
```diff
@@ -294,6 +295,33 @@ impl<'a> SuggestDao<'a> {
...
+                             FROM
+                                 mdn_custom_details
+                             WHERE
+                                 suggestion_id = :suggestion_id
+                            "#,
+                            named_params! {
+                                ":suggestion_id": suggestion_id
+                            },
+                            |row| {
+                                if full_suffix.starts_with(keyword_suffix) {
```
Comment (linabutler) [inline components/suggest/src/db.rs:313]: I wonder if we can do this check before we do the `query_row_and_then`? (I think we could do the same for AMO, but that can be a different PR; no need to do that here 😊) No need to query `mdn_custom_details` if the suffixes don't match.

File: components/suggest/src/db.rs [hunk 2]
```diff
@@ -583,6 +611,84 @@ impl<'a> SuggestDao<'a> {
...
+                named_params! {
+                    ":record_id": record_id.as_str(),
+                    ":title": suggestion.title,
+                    ":url": suggestion.url,
+                    ":score": suggestion.score,
+                },
+                |row| row.get(0),
+                true,
+            )?;
+            self.conn.execute(
```
Comment (linabutler) [inline components/suggest/src/db.rs:651]:
```suggestion
            self.conn.execute_cached(
```

**Review by linabutler — APPROVED** [review]
Thanks @ncloudioj!

Please be sure to trigger a [branch build](https://mozilla.github.io/application-services/book/howtos/branch-builds.html) for iOS and Android. Adding new suggestion types should be a backward-compatible change (I'm realizing it would be super helpful to document what's backward-compatible, and what's a breaking change 😅), but let's make sure there's no unexpected breakage.

File: components/suggest/src/db.rs (hunk 1 above)
Comment (ncloudioj) [reply components/suggest/src/db.rs:313]: Yup, nice catch! Let's do that.

File: components/suggest/src/db.rs (hunk 2 above)
Comment (ncloudioj) [reply components/suggest/src/db.rs:651]: Doh, copy&pasta 🙈, will update both.

File: components/suggest/src/store.rs [hunk 3]
```diff
@@ -2865,6 +2897,89 @@ mod tests {
                 []
                 "#]],
             ),
+            (
+                "keyword = prefix; MDN only",
```
[the remark this answers is not in the document]
Comment (ncloudioj) [reply components/suggest/src/store.rs:2901]: Absolutely, will do.

**Comment by ncloudioj** [comment]
Hey @linabutler, are you suggesting to follow this [particular step](https://mozilla.github.io/application-services/book/howtos/branch-builds.html#breaking-changes-in-an-application-services-branch) (i.e. for breaking changes) for this patch? If so, would adding `[firefox-android: 1876211-suggest-mdn]` to the PR title be sufficient?

**Comment by mhammond** [comment]
The branch-name in `[firefox-android: branch-name]` is a branch on the firefox-android repo. eg, if you knew this was a breaking change, you'd make a branch on that repo with the fixes and specify that name in this PR (I don't think it allows branches in forks, which is a shame). In this scenario where you assume there are no breaking changes, specifying `[firefox-android: main]` is probably what you want

**Comment by ncloudioj** [comment]
Gotcha! Thanks for the pointers, Mark!
